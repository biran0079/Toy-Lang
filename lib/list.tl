fun prefix(l, n){
  res = [];
  i=0;
  while(i<n){
    res += l[i];
    i += 1;
  }
  return res;
}
fun suffix(l,n){
  res=[];
  length = len(l);
  i=n;
  while(i<length){
    res += l[i];
    i += 1;
  }
  return res;
}


fun swap(a,i,j){
  t=a[i];
  a[i]=a[j];
  a[j]=t;
}

fun isSorted(l){
  length = len(l);
  i=1;
  while(i<length){
    if(l[i-1] > l[i])
      return 0;
    i=i+1;  
  }
  return 1;
}
