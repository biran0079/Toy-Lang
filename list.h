#ifndef _LIST_H_
#define _LIST_H_

typedef struct List {
  void** arr;
  int size, cap;
} List;

List* newList();
void listPush(List* lst, void* e);
void* listGet(List* lst, int idx);
void* listPop(List* lst); 
int listSize(List* lst);
void* listLast(List* lst);
void listSet(List* lst, int i, void* v);
List* listCopy(List* lst);
void listClear(List* lst);

#endif
