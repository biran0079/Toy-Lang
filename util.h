#ifndef _UTIL_H_
#define _UTIL_H_

#define MALLOC(T) (T *) tlMalloc(sizeof(T))
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <setjmp.h>
#include <time.h>
#include <stdarg.h>

void *tlMalloc(int size);
void *tlRealloc(void *t, int size);
void tlFree(void *t);
void *copy(void *t, int size);
char *copyStr(char *t);
char *catStr(char *s1, char *s2);
long strToLong(char *s);
char *literalStringToString(char *s);
int error(char *msg, ...);
char *getFolder(char *s);
FILE *openFromPath(char *s, char *mode);
char *readFileWithPath(char *path);
char *readFile(FILE *f);

#endif
