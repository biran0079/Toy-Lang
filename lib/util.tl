fun getRandomList(n){
  res=[];
  while(n>0){
    res += rand();  
    n = n - 1;
  }
  return res;
}


import list;

fun internalQsort(a, left, right){
  if(left>=right)return;
  i=left;
  j=left+1;
  k=j;
  while(j<=right){
    if(a[j]<a[i]) {
      list.swap(a,k,j);
      k=k+1;
    }
    j=j+1;  
  }
  k=k-1;
  list.swap(a, left, k);
  internalQsort(a,left,k-1);
  internalQsort(a,k+1,right);
}
fun qsort(l){
  internalQsort(l,0,len(l)-1);  
}


fun merge(l1, l2){
  res=[];
  len1 = len(l1);
  len2 = len(l2);
  i=0;
  j=0;
  while(i<len1 && j<len2){
    if(l1[i]<l2[j]){
      res += l1[i];  
      i += 1;
    }else{
      res += l2[j];
      j = j+1;
    }
  }
  while(i<len1) {
    res += l1[i];  
    i += 1;
  }
  while(j<len2) {
    res += l2[j];
    j += 1;
  }
  return res;
}
fun msort(lst){
  length = len(lst);
  if(length<=1)return;
  mid = length / 2;
  l1 = list.prefix(lst, mid);
  l2 = list.suffix(lst, mid);
  msort(l1);
  msort(l2);
  t = merge(l1, l2);
  i=0;
  while(i<length){
    lst[i] = t[i];
    i=i+1;  
  }
}

