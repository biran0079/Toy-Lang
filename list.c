#include "list.h"
#include "util.h"

static void resize(List *lst, int cap) {
  lst->arr = tlRealloc(lst->arr, cap * sizeof(void *));
  lst->cap = cap;
}

int newListC = 0, freeListC = 0;

#define INIT_LIST_CAP 1

List *newList() {
  newListC++;
  List *res = MALLOC(List);
  res->size = 0;
  res->cap = INIT_LIST_CAP;
  res->arr = tlMalloc(res->cap * sizeof(void *));
  return res;
}

void freeList(List *lst) {
  freeListC++;
  tlFree(lst->arr);
  lst->arr = 0;
  lst->cap = 0;
  lst->size = 0;
  tlFree(lst);
}

void listPush(List *lst, void *e) {
  if (lst->cap == lst->size) {
    resize(lst, lst->cap * 2);
  }
  lst->arr[lst->size++] = e;
}

void listPushFront(List *lst, void *e) {
  if (lst->cap == lst->size) {
    resize(lst, lst->cap * 2);
  }
  int i;
  for (i = lst->size; i > 0; i--) {
    lst->arr[i] = lst->arr[i - 1];
  }
  lst->arr[0] = e;
  lst->size++;
}

void *listPop(List *lst) {
  if (lst->size == 0) {
    error("cannot pop empty list\n");
  }
  return lst->arr[--lst->size];
}

void *listLast(List *lst) {
  assert(lst->size > 0);
  return lst->arr[lst->size - 1];
}

void *listSet(List *lst, int idx, void *v) {
  assert(idx < lst->size && idx >= 0);
  void *oldValue = lst->arr[idx];
  lst->arr[idx] = v;
  return oldValue;
}

List *listCopy(List *lst) {
  List *res = newList();
  int i, len = lst->size;
  for (i = 0; i < len; i++) listPush(res, listGet(lst, i));
  return res;
}

void listClear(List *lst) { lst->size = 0; }

static void msortInternal(void *a[], int l, int r, Comparator cmp, void *t[]) {
  if (r - l <= 1) return;
  int mid = (l + r) / 2;
  msortInternal(a, l, mid, cmp, t);
  msortInternal(a, mid, r, cmp, t);
  int k = l, i = l, j = mid;
  while (i < mid && j < r) {
    if (cmp(a[i], a[j]) <= 0)
      t[k++] = a[i++];
    else
      t[k++] = a[j++];
  }
  while (i < mid) t[k++] = a[i++];
  while (j < r) t[k++] = a[j++];
  for (i = l; i < r; i++) a[i] = t[i];
}

void listSort(List *lst, Comparator cmp) {
  void **t = tlMalloc(listSize(lst) * sizeof(void *));
  msortInternal(lst->arr, 0, lst->size, cmp, t);
  tlFree(t);
}

void listPopTo(List *lst, int size) {
  if (lst->size < size)
    error("size passed to listPopTo is less than list size\n");
  lst->size = size;
}

void increaseSizeTo(List *lst, int size) {
  if (size <= lst->size) return;
  int cap = lst->cap;
  if (cap < size) {
    while (cap < size) cap *= 2;
    resize(lst, cap);
  }
  memset(lst->arr + lst->size, 0, sizeof(void *) * (size - lst->size));
  lst->size = size;
}

void listSwap(List *lst, int i, int j) {
  if (i == j) return;
  void *t = listGet(lst, i);
  listSet(lst, i, listGet(lst, j));
  listSet(lst, j, t);
}
