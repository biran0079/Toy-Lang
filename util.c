#include "util.h"

void* copy(void* t, int size) {
  void* res = malloc(size);
  memcpy(res, t, size);
  return res;
}

char* copyStr(char* t) {
  return copy(t, strlen(t) + 1);
}

long strToLong(char* s){
  long res = 0;
  while(*s) {
    res*=10;
    res+=*s-'0';
    s++;
  }
  return res;
}

char* literalStringToString(char *s) {
  char *ss = malloc(strlen(s)+1);
  int l = 0;
  s++;  // skil "
  while(*s!='"') {
    if(*s == '\\') {
      s++;
      switch(*s){
        case 'n' : ss[l++] = '\n'; break;
        case 'r' : ss[l++] = '\r'; break;
        case '"' : ss[l++] = '"'; break;
        case '\\' : ss[l++] = '\\'; break;
        default:  ss[l++] = *s; break;
      }
    }else{
      ss[l++] = *s;
    }
    s++;
  }
  ss[l++] = 0;
  return realloc(ss, l);
}

void error(char* format,...) {
  va_list ap;
  va_start(ap, format);
  vfprintf(stderr, format, ap);
  va_end(ap);
  exit(-1);
}
