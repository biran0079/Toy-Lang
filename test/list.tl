print([1,2,3,4]);
a=[1,2,3];
print(a[0]);
print(a[1]);
print(a[2]);
a[0]=4;
print(a[0]);
print(a==[4,2,3]);
a[0]=a[0]+1;
print(a==[4,2,3]);
print(a==[5,2,3]);
