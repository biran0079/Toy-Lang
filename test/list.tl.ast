[stmts, [call, [id, print], [exps, [list, [int, 1], [int, 2], [int, 3], [int, 4]]]], [=, [id, a], [list, [int, 1], [int, 2], [int, 3]]], [call, [id, print], [exps, [list_access, [id, a], [int, 0]]]], [call, [id, print], [exps, [list_access, [id, a], [int, 1]]]], [call, [id, print], [exps, [list_access, [id, a], [int, 2]]]], [=, [list_access, [id, a], [int, 0]], [int, 4]], [call, [id, print], [exps, [list_access, [id, a], [int, 0]]]], [call, [id, print], [exps, [==, [id, a], [list, [int, 4], [int, 2], [int, 3]]]]], [=, [list_access, [id, a], [int, 0]], [+, [list_access, [id, a], [int, 0]], [int, 1]]], [call, [id, print], [exps, [==, [id, a], [list, [int, 4], [int, 2], [int, 3]]]]], [call, [id, print], [exps, [==, [id, a], [list, [int, 5], [int, 2], [int, 3]]]]], [call, [id, print], [exps, [list, [int, 1], [none], [none], [int, 3]]]]]
