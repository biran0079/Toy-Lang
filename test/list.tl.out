[1, 2, 3, 4]
1
2
3
4
1
0
1
[1, none, none, 3]
